### DISCUSSION TITLE
Allow middleware to cascade with rewrites and cookies

### DISCUSSION THREAD

**Discussion by ericmatthys** [vercel/next.js#36325] [Ideas]
Currently, middleware within a directory that relies on a rewrite will never run. Only redirects allow the middleware to cascade. For an app that relies heavily on rewrites, this means that there ends up being middleware functions with too much responsibility at or near the root.

```
pages/
├─ _middleware.ts
├─ country/
│  ├─ [country]/
│  │  ├─ index.tsx
│  │  ├─ _middleware.ts
```

Let's say `pages/_middleware.ts` rewrites to the `/country/[country]` directory based on geo ip.

```js
const url = req.nextUrl.clone();
const country = req.geo?.country ?? 'US';
url.pathname = `/country/${country}${url.pathname}`;

return NextResponse.rewrite(url);
```

The lack of a cascade after a rewrite means that `/country/[country]/_middleware.ts` will never run, making it impossible to break apart middleware and forcing all middleware logic to be in the root `_middleware.ts`. For example, an A/B test on a specific page like `/country/[country]/some/deeply/nested/page` would end up in the root `_middleware.ts` as opposed to alongside the page.

Another potential limitation to the cascade, assuming that middleware did run, is how to access cookies added to the response. Let's say the root `_middleware.ts` adds a user `uid` cookie to the response, and we want to access that value in a deeper `_middleware.ts`. Even without the rewrite, there doesn't seem to be a way of getting that value further down the cascade.

```ts
export function middleware(req) {
  const response = NextResponse.next();

  let userID = req.cookies[UID_COOKIE];

  if (!userID) {
    userID = crypto.randomUUID();
    response.cookie(UID_COOKIE, userID);
  }

  return response;
}
```

```ts
export function middleware(req) {
  const response = NextResponse.next();
  console.log(response.cookies);
  console.log(req.cookies);
  return response;
}
```

Both of those logged cookie objects will be empty for the first request. On the second request, only `req.cookies` will have the `uid` cookie. This would be another factor that prevents middleware from being broken apart into more maintainable functions.

**Comment by icyJoseph** [comment]
I now have to vouch for this. We have a top level middleware that makes sure certain query parameter is never present, among other things. We centralized all cases through a rewrite.

Now, we want to use nested middleware, but because of the top level rewrite, the downstream middleware doesn't get triggered. 

This solution does work though. Taken from [here](https://github.com/vercel/examples/blob/main/edge-functions/query-params-filter/pages/_middleware.ts).

```ts
import { NextRequest, NextResponse } from 'next/server'

const allowedParams = ['allowed']

export function middleware(req: NextRequest) {
  const url = req.nextUrl
  let changed = false

  url.searchParams.forEach((_, key) => {
    if (!allowedParams.includes(key)) {
      url.searchParams.delete(key)
      changed = true
    }
  })

  // Avoid infinite loop by only redirecting if the query
  // params were changed
  if (changed) {
    return NextResponse.redirect(url)
  }
 // otherwise do our rewrite logic
}
```

However, that, `changed` variable, does feel odd, and we now trigger a client side redirect.

**Comment by icyJoseph** [comment]
There has been an update: https://github.com/vercel/next.js/issues/36879#issuecomment-1125825500 and https://github.com/vercel/next.js/pull/36772

> This PR deprecates declaring a middleware under pages other than pages/_middleware. This is in the context of having a simpler execution model for middleware and also ships some refactor work. There is a ton of a code to be simplified after this deprecation but I think it is best to do it progressively.

**Reply by ericmatthys** [reply]
Thanks for sharing. In that case, I think this is settled.
